parity-clib/examples/cpp/main.cpp

```diff
@@ -14,172 +14,138 @@
-	"{\"method\":\"eth_subscribe\",\"params\":[\"newHeads\"],\"id\":1,\"jsonrpc\":\"2.0\"}"
+const std::vector<std::string> ws_subscriptions{
+    R"({"method":"parity_subscribe","params":["eth_getBalance",["0xcd2a3d9f938e13cd947ec05abc7fe734df8dd826","latest"]],"id":1,"jsonrpc":"2.0"})"s,
+    R"({"method":"parity_subscribe","params":["parity_netPeers"],"id":1,"jsonrpc":"2.0"})"s,
+    R"({"method":"eth_subscribe","params":["newHeads"],"id":1,"jsonrpc":"2.0"})"s,
 };
-
-// callback that gets invoked upon an event
-void callback(void* user_data, const char* response, size_t _len) {
-	Callback* cb = static_cast<Callback*>(user_data);
-	if (cb->type == CALLBACK_RPC) {
-		cb->counter -= 1;
-	} else if (cb->type == CALLBACK_WS) {
-		std::regex is_subscription ("\\{\"jsonrpc\":\"2.0\",\"result\":\"0[xX][a-fA-F0-9]{16}\",\"id\":1\\}");
-		if (std::regex_match(response, is_subscription) == true) {
-			cb->counter -= 1;
-		}
-	}
-}
+const std::vector<std::string> cli_args{"--no-ipc"s, "--jsonrpc-apis=all"s,
+                                        "--chain"s, "kovan"s};
+const std::vector<std::string> light_config{
+    "--no-ipc"s, "--light"s, "--jsonrpc-apis=all"s, "--chain"s, "kovan"s};
+} // namespace
 
 int main() {
-	// run full-client
-	{
-		std::vector<const char*> config = {"--no-ipc" , "--jsonrpc-apis=all", "--chain", "kovan"};
-		void* parity = parity_run(config);
-		if (parity_rpc_queries(parity)) {
-			printf("rpc_queries failed\r\n");
-			return 1;
-		}
-
-		if (parity_subscribe_to_websocket(parity)) {
-			printf("ws_queries failed\r\n");
-			return 1;
-		}
-
-		if (parity != nullptr) {
-			parity_destroy(parity);
-		}
-	}
-
-	// run light-client
-	{
-		std::vector<const char*> light_config = {"--no-ipc", "--light", "--jsonrpc-apis=all", "--chain", "kovan"};
-		void* parity = parity_run(light_config);
-
-		if (parity_rpc_queries(parity)) {
-			printf("rpc_queries failed\r\n");
-			return 1;
-		}
-
-		if (parity_subscribe_to_websocket(parity)) {
-			printf("ws_queries failed\r\n");
-			return 1;
-		}
-
-		if (parity != nullptr) {
-			parity_destroy(parity);
-		}
-	}
-	return 0;
+  using parity::ethereum::ParityEthereum;
+  // run full-client
+  {
+    ParityEthereum parity = parity_run(cli_args);
+    parity_rpc_queries(parity);
+    parity_subscribe_to_websocket(parity);
+  }
+
+  // run light-client
+  {
+    ParityEthereum parity = parity_run(light_config);
+    parity_rpc_queries(parity);
+    parity_subscribe_to_websocket(parity);
+    exit(1);
+  }
+  return 0;
 }
 
-int parity_rpc_queries(void* parity) {
-	if (!parity) {
-		return 1;
-	}
-
-	Callback cb { .type = CALLBACK_RPC, .counter = rpc_queries.size() };
-
-	for (auto query : rpc_queries) {
-		if (parity_rpc(parity, query.c_str(), query.length(), TIMEOUT_ONE_MIN_AS_MILLIS, callback, &cb) != 0) {
-			return 1;
-		}
-	}
-
-	while(cb.counter != 0);
-	return 0;
+namespace {
+void parity_rpc_queries(ParityEthereum &parity) {
```

Comment:
Doesn't have to be addressing in this PR: 

However, technically I think `cb.counter != 0` could end up with `UB` but since we are not running this in the CI anymore we could just those indefinitely or use a timer to short circuit it and exit with the error.